text/0000-dot-profile-utility-buildpack.md

```diff
@@ -0,0 +1,129 @@
+It will:
+
+- Detect a `.profile` file in the app dir
+- Wrap the `.profile` file so that it implements the `exec.d` interface
+- Add the `exec.d` executable to the `<layer>/exec.d` directory so the launcher will apply it
+
+We propose the ID `buildpacksio/profie` for this buildpack (cf. [RFC 97](https://github.com/buildpacks/rfcs/blob/main/text/0097-official-utility-buildpacks.md#what-it-is)).
+
+## Example 1 (environment variables)
+
+Here is an example of a `.profile` script, inspired by paketo-buildpacks/node-engine:
+
+```
+memory_in_bytes="$(cat /sys/fs/cgroup/memory/memory.limit_in_bytes)"
+MEMORY_AVAILABLE="$(( $memory_in_bytes / ( 1024 * 1024 ) ))"
+export MEMORY_AVAILABLE
+```
+
+The wrapper should ensure that the `MEMORY_AVAILABLE` environment variable is set in the environment with the proper value.
+
+## Example 2 (file system side effects)
+
+With this `.profile` script:
+
+```
+echo 'hello' >> "$HOME/hello"
+```
+
+The wrapper would not need to set any environment varibales, but should maintain the side effect of creating the `$HOME/hello` file.
+
+
+# How it Works
+[how-it-works]: #how-it-works
+
+In short, the wrapper should do:
+
+```
+#!/bin/bash
+set -euo pipefail
+source .profile
+env >&3
+```
+
+We acknowledge that the `env` output will need to be properly quoted to be valid TOML.
+We will use or create a tool to handle that formatting.
+
+This will write the environment, including any variables set in `.profile`, to the [`exec.d` output TOML](https://github.com/buildpacks/spec/blob/main/buildpack.md#execd-output-toml).
+Since it also executes the `.profile` script, any side effects will happen.
+This will solve for both of the simple examples above.
+
+On Windows, for `.profile.bat` scripts, we can take the same approach of wrapping the script like:
+
+```
+call .profile.bat
+set >&%CNB_EXEC_D_HANDLE%
+```
+
+# Migration
+[migration]: #migration
+
+This buildpack is new, so has no inherent migration considerations.
+
+For the related migration concerns for removing shell functionality in general, see the [Migration Path section of RFC 93](https://github.com/buildpacks/rfcs/blob/main/text/0093-remove-shell-processes.md#migration-path).
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+This is a new component to maintain and support.
+That being said, this will enable spec and lifecycle simplifications noted in [RFC 93](https://github.com/buildpacks/rfcs/blob/main/text/0093-remove-shell-processes.md) (where this approach was first suggested).
+Also, [RFC 97](https://github.com/buildpacks/rfcs/blob/main/text/0097-official-utility-buildpacks.md) resolves to support and maintain utility buildpacks.
+
+# Alternatives
+[alternatives]: #alternatives
+
+If we do nothing, we introduce a regression in functionality, and force application developers to rework their `.profile` scripts.
+
+
+# Prior Art
+[prior-art]: #prior-art
+
+-
+
+# Unresolved Questions
```

Comment:
Do we need to specify versions of Bash or shell that this will support?

Reply:
> Do we need to specify versions of Bash or shell that this will support?

Yes we should. The [spec](https://github.com/buildpacks/spec#operating-system-conventions) requires bash 3+ on linux, and cmd.exe (no specific version) on windows. Will add to the text.

Reply:
> How will this work if the user targets a `--run-image` that's FROM scratch (i.e. has no Bash installed). Should the exec.d error out, or should it just no-op?

It would seem less surprising to me if this buildpack would add a layer with bash.

Reply:
that's an interesting idea

Reply:
I did a little exploration with running bash on scratch. It was pretty straight-forward. Debian has a `bash-static` package. Extract the static bash binary from that, plop it down on a `FROM scratch` container, and Bob's your uncle.

Dockerfile and Makefile for this experiment here: https://github.com/mboldt/scratch-play/tree/main/bash

Details to figure out (keeping up with bash versioning, are we free license-wise to use the Debian package), but the add-layer-with-statically-linked-bash approach is feasible.

If the Debian package is any sort of blocker, there's also some prior art on compiling bash statically from source here: https://github.com/robxu9/bash-static

Reply:
That's great to know. fwiw, I don't think we need to answer all those questions in order to finish this RFC and ship the first version of the profile buildpack, but definitely something we will probably want in the future. 

Let's capture this info in the text of the RFC though.

Reply:
I would be in favor of declaring in the docs for the buildpack that bash is required in the run image, and if it's not present failing with a reasonable error message.

On one hand, it's convenient to install bash and make it work, on the other most folks are using an image w/out bash for the specific reason of not having bash. If we add bash, that defeats the purpose of using that image. I think the user should be presented with the decision, either use an image with bash or don't use `.profile` scripts that require bash.

This also keeps the buildpack implementation much simpler.

Reply:
yea, @dmikusa-pivotal makes a good point. It would be surprising to get a bash when you wanted scratch.

Maybe we can split the bash install into it's own buildpack, then it can be used as desired.

Reply:
Thank you for the discussion. The "bash buildpack" approach certainly sounds reasonable to me - more explicit and modular. I will capture this in the text.

Reply:
For what it's worth, I'm not sure we could accept a bash buildpack. The RFC for utility buildpacks in the requirements section says:

>The buildpack MUST NOT install any dependencies in the output image

https://github.com/buildpacks/rfcs/blob/main/text/0097-official-utility-buildpacks.md#how-it-works

I agree it seems like a useful buildpack. It could be provided through some other project though, like Paketo. Folks could still combine and use them together.

Reply:
I've updated the text to discourage the idea of a bash buildpack for the sake of .profile scripts.